File: models/Norway/SHAM/R/SHAM.cpp
// Fitting state-space models to seal populations with scarce data
// Tor Arne Øigård and Hans J. Skaug
// ICES Journal of Marine Science, 2014
// Contact:  Tor Arne Øigård  (toro@imr.no) and Hans Julius Skaug (skaug@imr.no)

#include <TMB.hpp>
#include <iostream>

//Bound between 0 and 1 
template <class Type>
Type ilogit(Type x){
  return Type(1.0)/(Type(1.0)+exp(-x));
}

//Bound between -1 and 1
template <class Type>
Type bound(Type x){
    return Type(2)/(Type(1) + exp(-Type(2) * x)) - Type(1);
}

//Logit Transform
template <class Type>
Type logitt(Type x){
  return log(x/(Type(1)-x));
}

//posfun-function
template <class Type>
Type posfun(Type x, Type eps){
 if (x>=eps) 
    {
      return x;
    }
    else
    {
     Type y=1.0-x/eps;
     return eps*(1./(1.+y+y*y));
    }
}

template<class Type>
Type objective_function<Type>::operator() ()
{
  //Input data
  DATA_INTEGER(Amax);                      //Maximum age group
  DATA_INTEGER(SSstart);                   //Start of state-space process
  DATA_ARRAY(Cdata);                       //Catch data
  DATA_INTEGER(Nc);                        //Number of years with catch data
  DATA_ARRAY(pup_production);              //Pup production estimates
  DATA_INTEGER(Np);                        //Number of years with pup production data
  DATA_ARRAY(Fecundity);                   //Observed fecundity rates
  DATA_INTEGER(Nf);                        //Number of years with fecundity data
  DATA_VECTOR(Pmat);                        //Birth ogives
  DATA_INTEGER(Npred);                     //Number of years to run projections
  DATA_ARRAY(priors);                      //Priors for parameters
  DATA_INTEGER(Npriors);                   //Number of priors
  DATA_INTEGER(xmax);                      //Length of state-space process
  DATA_VECTOR(CQuota);                     //Catch level in future projections
  DATA_VECTOR(cap);                        //Capelin biomass
  DATA_VECTOR(cod);                        //Cod biomass
  
 //Estimated parameters
  PARAMETER(logK);                         //Initial population size
  PARAMETER(Mtilde);                       //Natural adult mortality
  PARAMETER(M0tilde);                      //Natural pup mortality
  PARAMETER(ftilde);                       //Mean fecundity rate
  PARAMETER(atilde);                       //AR(1) parameter
  PARAMETER(b1tilde);
  PARAMETER(b2tilde);
  PARAMETER(logSigma);                     //sigma
  PARAMETER_VECTOR(u);                     //random effect vector
  
  //Transform estimated parameters
  Type K = exp(logK);
  Type M = ilogit(Mtilde);
  Type M0 = ilogit(M0tilde);
  Type fmean = ilogit(ftilde);
  Type a = bound(atilde);
  Type b1 = ilogit(b1tilde);
  Type b2 = ilogit(b2tilde);
  Type Sigma = exp(logSigma);
  
  //Adult and pup survival
  Type em = exp(-M);
  Type em0 = exp(-M0);

  array<Type>  Cmatrix(Nc+Npred+2,3);           //Catch data
  array<Type>  Pm(Nc+Npred+1,Amax);             //Birth ogive
  vector<Type> N0(Nc+Npred+2);                  //Pup abundance
  vector<Type> N1(Nc+Npred+2);                  //Abundance of one year and older seals
  vector<Type> Ft(Nc+Npred+2);                  //Fecundity rates
  vector<Type> logitFt(Nc+Npred+2);             //logit Fecundity rates
  vector<Type> b(Npriors);                 	//Concatination of parameters

  // Concatination of parameters into b
  b(0) = K;                                      // Initial population size
  b(1) = M;                                      // Natural adult mortality
  b(2) = M0;                                     // Natural pup mortality
  b(3) = fmean;                                  // Mean fecundity rate before state-space process        
  b(4) = a;                                      // AR 1 parameter
  b(5) = b1;
  b(6) = b2;
  b(7) = Sigma;                                  // Noise variance
 

  // Preliminary calculations - Preparations of Catch data
  Cmatrix(0,0) = 1945;
  Cmatrix(0,1) = 0;
  Cmatrix(0,2) = 0;
  for(int i=1;i<(Nc+1);i++){
    Cmatrix(i,0) = Cdata(i-1,0);
    Cmatrix(i,1) = Cdata(i-1,1);
    Cmatrix(i,2) = Cdata(i-1,2);
    }

  for(int i=(Nc+1);i<(Nc+Npred+2);i++){
    Cmatrix(i,0) = Cmatrix(i-1,0)+1;
    Cmatrix(i,1) = CQuota(0);
    Cmatrix(i,2) = CQuota(1);
    }
    
  // Preliminary calculations - Preparations of birth ogive
  // (version with constant curve)
  for(int i=0;i<(Nc);i++){
    for(int j=0;j<Amax;j++){
      Pm(i,j) = Pmat(j);
    }
  }
 
 for(int i=(Nc);i<(Nc+Npred+1);i++){
   for(int j=0;j<Amax;j++){
     Pm(i,j) = Pmat(j);
   }
 }
 
 
  // Prepare the fecundity-vector - hold fecundity constant up until year stmod
  // Logistic transformation to restrict to (0,1) - EQ 7
  Type counter = 0;
  for(int i=0;i<=(SSstart-Cdata(0,0));i++){
    Ft(CppAD::Integer(counter)) = fmean;
    logitFt(CppAD::Integer(counter)) = log((fmean+Type(1e-10))/(Type(1)+Type(1e-10)-fmean));  //Could use predefined functions from above instead....
    counter = counter + 1;
    }

  // Logistic transformation to restrict to (0,1) - EQ 7
  Ft(CppAD::Integer(counter)) = fmean;
  logitFt(CppAD::Integer(counter)) = log((fmean+Type(1e-10))/(Type(1)+Type(1e-10)-fmean)); 
  counter = counter + 1;

  Type logit_fmean = log((fmean+Type(1e-10))/(Type(1)+Type(1e-10)-fmean)); 
 
  // Initialize AR(1) process
  vector<Type> x(xmax);
  x(0) = 0;
 
  // AR(1)-process
  for(int i=1;i<xmax;i++)  
   {
     x(i) = a*x(i-1) + b1*cap(i) - b2*cod(i) + Sigma*u(i-1);		               // AR1 update
     logitFt(CppAD::Integer(counter)) = logit_fmean+x(i);      // Part of Eq 7
     Ft(CppAD::Integer(counter)) = exp(logitFt(CppAD::Integer(counter)))/(Type(1)+exp(logitFt(CppAD::Integer(counter))));  // Logistic transformation to restrict to (0,1) - EQ 7
     counter = counter + 1;
   }


  // Initiate of N in year 0 (1945) - EQ 1 and 2
  matrix<Type> N(Nc+Npred+2,Amax);
  for(int i=0;i<Amax;i++){
    N(0,i) = exp(-Type(i)*M);		// Adults
  }

  N(0,Amax-1) /= 1-em;			// Correct A+ group
 
  Type Nsum = 0;
  for(int i = 0;i<Amax;i++)
    Nsum = Nsum + N(0,i);
  
  for(int i = 0;i<Amax;i++)
  N(0,i) = K*N(0,i)/Nsum;		// Normalize vector to K
  
  N1(0) = K;                            // Abundance of 1 year and older seals
  N0(0) = (1-em)/em0*K;			// To balance natural mortality of 1+ group

  // Calculate population trajectory
  for(int i=1;i<Nc+Npred+2;i++)
  {
    N(i,0) = (N0(i-1)-Cmatrix(i-1,1))*em0;			// 0-group from last year - EQ 3
    for(int j=1;j<Amax;j++)
    {
      N(i,j) = N(i-1,j-1)*(Type(1)-Cmatrix(i-1,2)/N1(i-1))*em;	// Pro-rata distribution of catches EQ 3 and 4
    }  

    N(i,Amax-1) += N(i-1,Amax-1)*(Type(1)-Cmatrix(i-1,2)/N1(i-1))*em;      // A+ group correction - EQ 3

    // Ensures that N > 0
    Nsum = 0;
    for(int j = 0;j<Amax;j++)
    Nsum = Nsum + N(i,j);

    N1(i) = posfun(Nsum-Cmatrix(i,2),Type(1)) + Cmatrix(i,2);
    for(int j = 0;j<Amax;j++)
    N(i,j) = N1(i)*N(i,j)/Nsum;			// Scales up the full age structure
  
    // Reqruitement equation
    for(int j=0;j<Amax;j++)
      N0(i) += 0.5*Ft(i)*Pm(i-1,j)*N(i,j);
   

    // Ensures that pup production is larger than pup catch
    N0(i) = posfun(N0(i)-Cmatrix(i,1),Type(1)) + Cmatrix(i,1);	
  }


  // Likelihood contribution
  Type nll=Type(0);

  //Data contribution to likelihood
  
  // Likelihood contribution from pup production estimates - EQ 9// Pup production estimates
  for(int i=0;i< Np;i++)
  {
    nll += -dnorm(N0(CppAD::Integer(pup_production(i,0)-Cdata(0,0)+1)),pup_production(i,1),pup_production(i,1)*pup_production(i,2),true);
  }

  // Likelihood contribution from fecundity estimates - EQ 10
  for(int i=0;i<Nf;i++)
  {
    nll += -dnorm(Ft(CppAD::Integer(Fecundity(i,0)-Cdata(0,0)+1)),Fecundity(i,1),Fecundity(i,2),true);
  }

   // Likelihood contribution from prior distibutions - EQ 11
  for(int i=0;i< Npriors;i++)
  {
    nll += -dnorm(b(i),priors(i,0),priors(i,1),true);
  }
  
  // Random effect contribution
  for(int i=0;i<xmax-1;i++)
  {
    nll += -dnorm(u(i),Type(0),Type(1),true);
  }

  ADREPORT(N0)
  ADREPORT(N1)
  ADREPORT(Ft)

  return nll;
}

File: models/Norway/SHAM/R/SHAM.R
## Including capelin and cod biomass as covariates as covariates on fecundity 
## Martin Biuw and Tor Arne Øigård
## Modified from:
## Fitting state-space models to seal populations with scarce data
## Tor Arne Øigård and Hans J. Skaug
## ICES Journal of Marine Science, 2014
## Contact:  Martin Biuw (martin.biuw@hi.no) or Tor Arne Øigård (tor.arne.oigard@nr.no)
source('./R/read.ICES.R')


library(TMB)

#####################################
#Some functions needed below
ilogit1 <- function(x){
  return(1/(1+exp(-x)))
}

logit <- function(x){
	return(log(x/(1-x)))
}

ilogit <- function(x){
	return(exp(x)/(1+exp(x)))
}
######################################

compile("R/SHAM.cpp")
dyn.load(dynlib("R/SHAM"))

###########
#Data part
###########
#Read in data
catch_data <- read.table("./Data/harpeast/catch_data.dat",header = FALSE)				#Catch data
pup_production <- read.table("./Data/harpeast/pup_production.dat",header = FALSE)			#Pup production estimates
fecundity <- read.table("./Data/harpeast/fecundity.dat",header = FALSE)				#Available fecundity data
#Pmatrix <- read.table("hessm.pma",header = FALSE)				#Birth ogives
#With collapsed maturity curves
Pmatrix <- read.table("./Data/harpeast/collapsed.ogi",header = TRUE)				#Birth ogives
Pmatrix <- c(Pmatrix[,2], rep(1, 4))
priors <- read.table("./Data/harpeast/priors.dat",header = FALSE)					#Priors used
years.of.prediction = 15											#Number of years to run projections

## Read and prepare fish data
cap <- read.ICES()
## Read capelin index from cod stomach contents 
## prior to 1972 (From Marshall et al. 2000)
histCap <- read.csv('./Data/histCap.csv')
histCap$method <- rep('Rec', nrow(histCap))
cap$Data <- plyr::rbind.fill(histCap[-nrow(histCap),], cap$Data)
cap$Data$method[which(is.na(cap$Data$method))] <- 'Meas'
cap$Data$method <- as.factor(cap$Data$method)
cod <- read.ICES('./Data/_9841_9841.xml')

fish <- cod$Data[,c(1,6,9)]
names(fish)[c(2,3)] <- c('CodTB', 'CodSSB')
fish <- merge(fish, cap$Data[,c(1:4)], by='Year', all.x=T, all.y=F)
names(fish)[c(4,6)] <- c('CapTB', 'CapSSB')
fish <- fish[which(!is.na(fish$CodTB) & !is.na(fish$CapTB)),]
fish$suit <- (fish$CapTB/max(fish$CapTB, na.rm=T))-(fish$CodTB/max(fish$CodTB, na.rm=T))
fish$suit <- fish$suit-min(fish$suit)
fish$suit <- fish$suit/max(fish$suit)


#Prepare input to the model
data <- list()
data$Amax = 20													#Maximum age group
data$SSstart = 1985												#State-space process starts from this year
data$Cdata = as.matrix(catch_data)								#Add catch data
data$Nc = length(catch_data[,1])									#Number of years with catch data
data$pup_production = as.matrix(pup_production)					#Pup production estimates
data$Np = length(pup_production[,1])								#Number of years with pup production data
data$Fecundity = as.matrix(fecundity)							#Observed fecundity rates
data$Nf = length(fecundity$V1)									#Number of years with fecundity data
data$Pmat = as.matrix(Pmatrix)									#Birt ogives
data$Npred = years.of.prediction									#Number of years to run projections
data$priors = as.matrix(priors)									#Priors for estimated parameters
data$Npriors = length(priors$V1)									#Number of priors
data$xmax = data$Cdata[data$Nc,1]+data$Npred+1-data$SSstart+1	#Length of state-space process
data$CQuota = c(0,0)											#Catch level in future projections
data$cap = fish$CapTB/max(fish$CapTB)              #Capelin biomass
data$cod = fish$CodTB/max(fish$CodTB)              #Cod biomass

###############
#Parameter part
###############

#Initial values
Kinit = 2000000								#Initial population size
Minit = 0.09									#Natural adult mortality
M0init = 0.27								#Natural pup mortality
finit = 0.7									#Mean fecundity rate
ainit = 0.7									#AR(1) parameter
b1init = 0.4
b2init = 0.6
sigmainit = 0.85								#sigma

#Transform some parameters to ensure that estimates are > 0
parameters <- list()
parameters$logK= log(Kinit)  			
parameters$Mtilde= logit(Minit)				
parameters$M0tilde= logit(M0init)				
parameters$ftilde = logit(finit)
parameters$atilde = ainit
parameters$b1tilde = logit(b1init)
parameters$b2tilde = logit(b2init)
parameters$logSigma = log(sigmainit)
parameters$u=rep(0,length.out = (data$Nc+data$Npred-data$SSstart+data$Cdata[1,1]))


obj <- MakeADFun(data,parameters,random="u",DLL="SHAM")

obj$fn()
obj$gr()

system.time(opt<-nlminb(obj$par,obj$fn,obj$gr))

rep<-sdreport(obj, getJointPrecision=TRUE)
rep.matrix <- summary(rep)
rep.rnames = rownames(rep.matrix)
indN0 = which(rep.rnames=="N0");indN0 <- indN0[-1]
indN1 = which(rep.rnames=="N1");indN1 <- indN1[-1]
indFt = which(rep.rnames=="Ft");indFt <- indFt[-1]
yrs = 1946:2035

N0 =rep.matrix[indN0,1]
N0sd = rep.matrix[indN0,2]
N1 = rep.matrix[indN1,1]
N1sd = rep.matrix[indN1,2]
Ft = rep.matrix[indFt,1]
Ftsd = rep.matrix[indFt,2]

#Extract parameters
Kest = exp(opt$par[1])          
Mest = ilogit(opt$par[2])       
M0est = ilogit(opt$par[3])      
fest = ilogit(opt$par[4])       
aest = opt$par[5]
b1est = ilogit(opt$par[6])
b2est = ilogit(opt$par[7])
sigmaest = exp(opt$par[8])    

save_results = FALSE
if(save_results){
#Save results
#Original model
#filename = "OriginalStateSpaceModel.Rdat"
#With prey data
  filename = "NewStateSpaceModel.Rdat"
  save(Kest = Kest,Mest = Mest,M0est=M0est,fest = fest,aest = aest,b1est = b1est,b2est=b2est,sigmaest = sigmaest,N1 = N1,N1sd = N1sd,N0 = N0,N0sd = N0sd,Ft = Ft,Ftsd = Ftsd,file = filename)
}
#allsd<-sqrt(diag(solve(rep$jointPrecision)))
#plsd <- obj$env$parList(par=allsd)
# X11("",9,7)
# plot(1946:2035,rep.matrix[indN1,1],type = "l",xlab = "Year",ylab = "Abundance",xlim = c(1946,2030),ylim = c(0,2000000),lwd = 3,col = "darkgreen")
# lines(1946:2035,rep.matrix[indN0,1],col = "blue",lwd = 3)
# lines(data$pup_production[,1],data$pup_production[,2],type = "p",col = "red",cex = 1.2,pch = 16)

#################
#Plot figures

#load("Data/Results/NewStateSpaceModel.Rdat")
dfnew=data.frame(Year=1946:2035,
              N1new=N1,
              N0new=N0,
              N1newl=N1-1.96*N1sd,
              N1newu=N1+1.96*N1sd,
              N0newl=N0-1.96*N0sd,
              N0newu=N0+1.96*N0sd,
              Fnew=Ft,
              Fnewl=Ft-1.96*Ftsd,
              Fnewu=F+1.96*Ftsd)

load("./Data/Results/OriginalStateSpaceModel.rdat")
dfold=data.frame(Year=1946:2035,
                 N1old=N1,
                 N0old=N0,
                 N1oldl=N1-1.96*N1sd,
                 N1oldu=N1+1.96*N1sd,
                 N0oldl=N0-1.96*N0sd,
                 N0oldu=N0+1.96*N0sd,
                 Fold=Ft,
                 Foldl=Ft-1.96*Ftsd,
                 Foldu=F+1.96*Ftsd)

df = data.frame(dfnew,dfold)

library(ggplot2)
theme_set(theme_minimal())

#Pup abundance
X11("",6,4)
ggplot(data=df,aes(x=Year))+
  geom_line(aes(y=N0new),colour="royalblue",size = 1.5)+
  geom_ribbon(aes(y = N0new,ymin = N0newl, ymax = N0newu),fill='lightblue',alpha=0.5)+
  #geom_line(aes(y=N0old),colour='red',size=1.5)+
  #geom_ribbon(aes(y = N0old,ymin = N0oldl, ymax = N0oldu),fill='lightpink',alpha=0.5)+
  labs(x='Year',y='Pup abundance',title='')

#Fecundity
X11("",9,7)
ggplot(data=df,aes(x=Year))+
  geom_line(aes(y=Fnew),colour="royalblue",size = 1.5)+
  #geom_ribbon(aes(y = Fnew,ymin = Fnewl, ymax = Fnewu),fill='lightblue',alpha=0.5)+
  #geom_line(aes(y=Fold),colour='red',size = 1.5)+
  #geom_ribbon(aes(y = Fold,ymin = Foldl, ymax = Foldu),fill='lightpink',alpha=0.5)+
  labs(x='Year',y='Modelled fecundity',title='')